src/splay_tree.jl

```diff
@@ -137,33 +137,31 @@ function search_node(tree::SplayTree{K}, d::K) where K
             node = node.leftChild
         end
     end
-    return (node == nothing) ? prev : node
+    last = (node == nothing) ? prev : node
+    (last == nothing) || splay!(tree, last)
+    return last
 end
 
 function Base.haskey(tree::SplayTree{K}, d::K) where K
     node = tree.root
     if node === nothing
         return false
     else
-        node = search_node(tree, d)
+        node = search_node!(tree, d)
         (node === nothing) && return false
-        is_found = (node.data == d)
-        is_found && splay!(tree, node)
-        return is_found
+        return (node.data == d)
     end
 end
 
 Base.in(key, tree::SplayTree) = haskey(tree, key)
```

Comment:
Inline with my above comment, every function calling `search_node!` becomes modifying in nature.

Reply:
This becomes moot, because of "user-visible state" definition.